scalableminds/webknossos
# Zarr3 download for volume annotation, supporting nd

### URL of deployed dev instance (used for testing):
- https://ndvolumedownload.webknossos.xyz

### Steps to test:
- Create annotation on 3d dataset, see download modal, try different options
- Create annotation on nd dataset, see download modal, try different options
- Download from other places like dashboard should still use wkw for the moment (untill the libs support zarr)
- Reupload of a zarr volume annotation should yield the same annotation
- Zarr metadata and datasource-properties.json in data.zip should look right.
- Unpacking a zarr data.zip and moving it to the a folder in binaryData should yield a usable webknossos dataset (this does not yet work in nd case because only specific axis orders are supported with additional coordinates)

### TODOs:
  - [x] add parameter selecting download as zarr
  - [x] write zarr chunks
  - [x] support nd in writing zarr chunks, (what axis order? should channels be included?)
  - [x] write zarr header (how too persist axis order?)
    - [x] Q: should the shape be the full layer shape? (starting at 0,0,0)? A: yes
    - [x] Q: should the fill_value be set? (0 is kind of wrong due to fallback layer?) A: use 0 anyway
    - [x] Q: should we set any additional attributes?
    - [ ] Get rid of the auto-generated field for the codecs → I don’t know how :(
  - [x] save compressed buckets
  - [x] save compression info in the zarr.json
  - [x] make parameter enum instead
  - [x] include dataZipFormat as tag in NML
  - [x] support upload
    - [x] in upload, distinguish between volume formats
       - [x] Q: should this be auto-detected from the contents or taken from the NML? Auto-detect for now
    - [x] get bucket position with additional coordinates from paths
    - [x] in upload, include assertions for the right zarr format
    - [x] support nd
    - [x] support merging volumes when uploading multiple
  - [x] include datasource-properties.json in data.zip to enable using this as a normal dataset
  - [x] include channel axis
    - [x] why does wk no longer display the layer as normal dataset? (even in 3d+c case)
  - [x] in zarr.json metadata, use array-of-int transpose codec instead of "F".
  - [x] Select zarr3 download format for annotations with additional coordinates
  - [x] Say so in the download modal
  - [x] In places where downloads are started without the modal (i.e. dashboard), keep wkw for now until the libs support the zarr-based format, compare https://github.com/scalableminds/webknossos-libs/issues/951

### Issues:
- fixes #7273
- contributes to  #7116

------
